ZcashFoundation/zebra
# fix(network): Add a send timeout to outbound peer messages

## Motivation

Zebra doesn't impose a timeout when sending messages to the remote peer. (It only has one when receiving them.)

So at the moment, we're relying on the 2 minute disconnection timeout from the OS TCP stack, and Zebra's internal 2 minute peer service disconnection timeout. (If they actually work, which we haven't tested.)

## Solution

Production Code:
- wrap the peer message sender in a type that adds a timeout
- add a `Connection::new` method

Tests:
- use bounded fake peer connection channels in the connection tests, so we can actually trigger send timeouts 
- test peer send and receive timeouts 
- check for specific error types in all the connection test vectors
- add docs to all the connection test vectors

Closes #3234
Closes #3233 
Closes #3232

## Review

@oxarbitrage can review this PR, he is familiar with ticket #3234.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors
